[MRG+1] PY3 exporters

eliasdorneles opened the pull request:
Hey fellows,

I've started porting the exporters to Python 3, but I'm not sure how backward compatible we should be -- I'm creating the PR already hoping it'll help moving this forward.

There isn't much here yet, I just got the first few tests passing under Python 3, keeping the retrocompatibility. Current passings tests are:

```
tox -e py27,py34 -- tests/test_exporters.py::BaseItemExporterTest tests/test_exporters.py::PythonItemExporterTest tests/test_exporters.py::CustomItemExporterTest
```

According to issue #1080 (which I tend to agree), we should change the default output of BaseItemExporter to unicode text and only do encode to bytes for things that we can't use text (like CSV writer in Python 2 only). This would make porting easier, but would cause issues for users since we'd be changing the default file format/encoding.

The JSON encoder even rejects bytes in Python 3, trying to `json.dumps(b'hi')` gets you `TypeError: b'hi' is not JSON serializable`.

The CSV writer is particularly tricky because in PY2 it expects a binary stream to write and in PY3 it expects a text stream. The [Python 3 Porting guide proposes an interesting solution](http://python3porting.com/problems.html#csv-api-changes) which requires control of file opening, but our Item Exporters API gets a file handle.
To make matters more complicated, the FileFeedStorage in feed exports extension opens the file always in binary mode: https://github.com/scrapy/scrapy/blob/master/scrapy/extensions/feedexport.py#L79

I'd appreciate some guidance! :)

eliasdorneles commented:
Some alternatives that I see:

1) Move to export in UTF-8 and accept the break in compatibility (since it is arguably a better default behavior).

2) Keep the current behavior for Python 2, and favor exporting in UTF-8 only in Python 3. This is a bit messier, but would already make some things easier and might be an acceptable tradeoff.

3) Continue the path I'm taking here, trying to make it backward compatible and working around the stdlib modules that doesn't accept bytes. It's getting a bit ugly, it would be great to avoid it, but I dunno.

Thoughts?

kmike commented:
Hey,

I think that moving JSON and XML exporters to unicode is a backwards compatible change, CSV should be unicode only in Python 3.x, Pickle and Marshal move to unicode is a bug fix, and the only backwards incompatible change is for PythonItemExporter.

For PythonItemExporter we can add an option (`binary=True`) to encode data to bytes and make it default, but issue a warning if default value is used. Or just break backwards compatibility for PythonItemExporter.

So my vote is for (1). To clarify, we shouldn't move to export in UTF-8; our exporters should use unicode and pass data to library, and try to avoid encoding data to any specific encoding (like UTF-8). In Python utf-8 and unicode are antonyms, not synonims :)

eliasdorneles commented:
that's not true, Github, I haven't actually closed -- I'm just restarting from scratch. :P

eliasdorneles commented:
Upon some reflection on why I got stuck with this, I realized it was because there is a good bit of coupling not only in the production code but also in the tests.

So, I restarted from scratch, doing it in smaller steps now: I've changed BaseItemExporter and its test to the behavior we want (use unicode objects and avoid decoding to UTF-8), then pushed down the test methods that needed changing to a new "intermediary" base test class serving the base for the test classes for the others that need the previous behavior.

It sounds a bit complicated, but the idea is simple: do the smallest possible change at each time, while keeping the test suite still working. I'm slightly more confident about this approach, idea is to change the behavior according #1080 and only then port to PY3.

kmike commented on scrapy/exporters.py:
```diff
@@ -139,8 +139,9 @@ def _export_xml_field(self, name, serialized_value):
-            for value in serialized_value:
+        elif (hasattr(serialized_value, '__iter__')
+              and not isinstance(serialized_value, six.string_types)):
+              for value in serialized_value:
```
Is it possible to use https://github.com/scrapy/scrapy/blob/e0074d7f4d3a7fcff3f3f00b4db7175b7fdf2fba/scrapy/utils/python.py#L47 here?

Also, indentation looks off :)

eliasdorneles replied:
aw yea, much better, I had forgotten about that one -- thanks! :)

eliasdorneles commented:
So, today I got together with @stummjr for a day sprint wrapping up the work for this port.

We followed the advice from @kmike on the above comments and in #1080 and we were finally able to run scrapy with Python 3 exporting items to JSON, XML and CSV -- so nice to see this working! =)

We weren't sure of some of the decisions, looking forward to your comments.

Thank you!

kmike commented on scrapy/exporters.py:
```diff
@@ -249,13 +263,20 @@ def _serialize_value(self, value):
 
     def _serialize_dict(self, value):
         for key, val in six.iteritems(value):
+            key = to_bytes(key) if self.binary else key
```
Is this method called when `self.binary` is False?

eliasdorneles replied:
you mean, `_serialize_dict`?

yes, it's called for nested items in `_serialize_value` (function above).

it has an extra call in `_get_serialized_fields` to handle the top-level item keys. ;-)

kmike commented on scrapy/exporters.py:
```diff
@@ -239,6 +246,13 @@ class PythonItemExporter(BaseItemExporter):
+        if self.binary:
+            warnings.warn(
+                "PythonItemExporter will drop support for binary export in the future",
+                PendingDeprecationWarning)
```
Why not DeprecationWarning?

eliasdorneles replied:
I wasn't sure, it wasn't quite clear to me how to decide between the two:

```
exception DeprecationWarning
Base class for warnings about deprecated features.

exception PendingDeprecationWarning
Base class for warnings about features which will be deprecated in the future.
```

I'll change it to DeprecationWarning. :+1:

kmike commented on tests/test_exporters.py:
```diff
@@ -267,19 +276,19 @@ def test_nested_item(self):
-                    '<name>buz</name>'
-                '</item>'
-            '</items>'
+            u'<?xml version="1.0" ?>\n'
```
hm, so it no longer sets XML encoding declaration? It looks like a regression

eliasdorneles replied:
So, this is just setting up the expectation, the thing is that lxml.etree.tostring complains if you give it a text object with `encoding="utf-8"` (it was needed before because the expected value was defined in bytes).

They're "XML equivalent":

```
$ python
Python 3.4.3 (default, May 27 2015, 20:22:45) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lxml.etree
>>> b = b'<?xml version="1.0" encoding="utf-8"?>\n<items>hello</items>'
>>> u = u'<?xml version="1.0"?>\n<items>hello</items>'
>>> def xmltuple(elem):
...     children = list(elem.iterchildren())
...     if children:
...         return [(child.tag, sorted(xmltuple(child)))
...                 for child in children]
...     else:
...         return [(elem.tag, [(elem.text, ())])]
... 
>>> xmltuple(lxml.etree.fromstring(b)) == xmltuple(lxml.etree.fromstring(u))
True
>>> 
```

So, we could make the expectation bytes and keep the `encoding="utf-8"` and it would be equivalent, what do you think?

kmike commented:
Overall it looks good!

Would you mind adding a couple more FeedExporter tests? Currently we only check csv and json lines, it'd be good to test that other exporters work with feed exports.

eliasdorneles commented:
Sure, I'll add some other tests to feed exports too. :+1:

eliasdorneles commented:
@kmike I've addressed your comments, and added a test for feed export XML output.
I tried adding tests for pickle and marshal output, but I think they're broken, hence I created issue #1713

kmike commented:
:+1:

redapple commented on scrapy/exporters.py:
```diff
@@ -86,10 +88,10 @@ def __init__(self, file, **kwargs):
 
 
-class JsonItemExporter(JsonLinesItemExporter):
+class JsonItemExporter(BaseItemExporter):
```
`JsonItemExporter` and `JsonLinesItemExporter` still share common stuff.
what's the reason for having them both subclass `BaseItemExporter`?

eliasdorneles replied:
This is by design, `BaseItemExporter` implements the stuff common to all exporters (recursively export fields, fields_to_export and export_empty_fields settings).

kmike replied:
I think it makes more sense to have them both as subclasses of `BaseItemExporter` than inheriting `JsonItemExporter` from `JsonLinesItemExporter` and overriding *all* methods. It is also not is-a relationship. There is a bit of code duplication in `export_item` and `__init__`, but it's not that bad.

redapple replied:
hm, ok. I was imagining an exporter class that can do both, and having a `stream` constructor argument of some sort to do lines or not.
no big deal.

dangra commented:
`scrapy/exporters.py` [is near 100% code coverage](https://codecov.io/github/scrapy/scrapy/scrapy/exporters.py?ref=d515212209051f48ad7cb634d8a02b2ed5efc50d). Is it asking too much to add test cases for missing code lines?

eliasdorneles commented:
@dangra so, I improved the coverage a tiny little bit, but 100% of coverage because of reasons I cover below.

While improving coverage, I realized that the [except TypeError: pass in CsvItemExporter](https://codecov.io/github/scrapy/scrapy/scrapy/exporters.py?ref=d515212209051f48ad7cb634d8a02b2ed5efc50d#l-180) had kinda undefined behavior -- it's for the case when transforming lists to CSV which aren't made of strings/bytes. The current behavior (code in master, traced back to [this commit](https://github.com/scrapy/scrapy/commit/8a5c08a6bcde96938ed92ee4996aafb6655c3688)) is to just pass lists of non-strings to [csv.writer.writerow](https://docs.python.org/2/library/csv.html#csv.csvwriter.writerow), which would put in the CSV the Python representation of the object.
So I updated the new code (which was crashing trying to pass lists of non-strings to `to_native_str`), making explicit the fallback to `repr()` and added a test for it.

So, the reasons why getting to 100% coverage would be just awkward are:

1) `MarshalItemExporter` is already covered by an integration test in `test_feedexport.py` that runs a spider and exports to an actual file (btw, why isn't that showing up in the coverage?). The marshal module is awkward to test isolatedly in PY 2 because it doesn't work with file-like objects, only with actual files:

```
>>> import marshal, io
>>> marshal.dump({}, io.BytesIO())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-4864edcc4e37> in <module>()
----> 1 marshal.dump({}, io.BytesIO())

TypeError: marshal.dump() 2nd arg must be file
```

2) The [workaround in XmlItemExporter](https://codecov.io/github/scrapy/scrapy/scrapy/exporters.py?ref=d515212209051f48ad7cb634d8a02b2ed5efc50d#l-151) is unreachable unless it runs in Python <= 2.7.3 -- which we're not running it in Travis, so I don't really see the point.

So, does this look good?

dangra commented:
@eliasdorneles: I added a test case for MarshalItemExporter and a coverage exclusion for XmlItemExporter workaround. Interestingly it is not 100% due to partial coverage at https://codecov.io/github/scrapy/scrapy/scrapy/exporters.py?ref=6fc286753faadb22db89ea66e0ef934c0840ec29#l-275

eliasdorneles commented:
awesome! \o/

lopuhin commented:
py3 jumped to 94%, yay!

dangra commented:
> wouldn't we want to clean up the temp file?

from `tempfile.TemporaryFile()` docs: 

> The file is created using mkstemp(). It will be destroyed as soon as it is closed (including an implicit close when the object is garbage collected).

eliasdorneles commented:
ahh, muy bien! :+1: :+1: =)
